projects/motor_controller_interface/src/motor_controller_fsm.c

```diff
@@ -0,0 +1,322 @@
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_reverse_guard, torque_control_reverse);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_cruise_control_forward_guard, cruise_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, prv_mechanical_braking_guard,
+                             mechanical_braking);
+}
+
+FSM_STATE_TRANSITION(mechanical_braking) {
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_forward_guard, torque_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_reverse_guard, torque_control_reverse);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_cruise_control_forward_guard, cruise_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, prv_mechanical_braking_guard,
+                             mechanical_braking);
+}
+
+// Output functions
+static void prv_torque_control_mode(FSM *fsm, const Event *e, void *context) {
+  DriverControlsData data = { 0 };
+  fifo_pop(s_fifo, &data);
+
+  float desired_torque = CONVERT_THROTTLE_READING_TO_PERCENTAGE((float)data.throttle);
+
+  if (data.brake_state == DRIVER_CONTROLS_BRAKE_ENGAGED) {
+    // If the brake is enaged, then ensure that we send 0 torque
+    desired_torque = 0;
+  }
+
+  float desired_direction = 0;
+  if (data.direction == DRIVER_CONTROLS_REVERSE) {
+    desired_direction = MOTOR_CONTROLLER_REVERSE_VELOCITY_MPS;
+  } else {
+    // In case we receive an invalid direction, it's probably safer to go forward
+    // rather than backwards
+    desired_direction = MOTOR_CONTROLLER_FORWARD_VELOCITY_MPS;
+  }
+
+  // To run the motor in torque control mode, set the velocity to an
+  // unobtainable value such as 100 m/s. Set the current to a value that is
+  // proportional to your accelerator pedal position.
+  MotorControllerDriveCmd cmd = {
+    .current_percentage = desired_torque,
+    .velocity = desired_direction,
+  };
+
+  // Update Left Motor Controller message
+  GenericCanMsg left_msg = {
+    .id = MOTOR_CONTROLLER_DRIVE_COMMAND_ID(MOTOR_CONTROLLER_INTERFACE_LEFT_ADDR),
+    .data = 0,
+    .dlc = sizeof(cmd),
+    .extended = false,
+  };
+  memcpy(&left_msg.data, &cmd, sizeof(cmd));
+  s_left_interval->msg = left_msg;
+
+  // Update Right Motor Controller message
+  GenericCanMsg right_msg = {
+    .id = MOTOR_CONTROLLER_DRIVE_COMMAND_ID(MOTOR_CONTROLLER_INTERFACE_RIGHT_ADDR),
+    .data = 0,
+    .dlc = sizeof(cmd),
+    .extended = false,
+  };
+  memcpy(&right_msg.data, &cmd, sizeof(cmd));
+  s_right_interval->msg = right_msg;
+
+  // Trigger a transmission
+  can_interval_send_now(s_left_interval);
```

Comment:
Would it be possible to use array-based intervals/messages instead of managing two independent variables?

Reply:
You mean making `s_right_interval` and `s_left_interval` as an array instead? Something like

```c
#define NUM_CAN_INTERVALS 2
static CanInterval * s_can_intervals[NUM_CAN_INTERVALS] = { 0 };

// do stuff to left
can_interval_send_now(s_can_intervals[0]);

// do stuff to right
can_interval_send_now(s_can_intervals[1]);
```

Not sure if that's what you meant

Reply:
Yeah, except you can generate the message in a loop as well. Something like

```
for (size_t i = 0; i < NUM_CAN_INTERVALS; i++) {
  GenericCanMsg msg = { ... };
  // set can interval message
  // trigger transmission
}
```